### PR TITLE
refactor: Refactor navigation

### PR DESCRIPTION
**Pull request by andiwand**
After #2625 turned out to be harder than initially thought I tried to factor out some changes to get them into main already.

Summary of the changes I made
- `pLimit` = path limit -> far limit = `farLimit`
- `oLimit` = overstep limit -> near limit = `nearLimit`
- replaced `DetectorNavigator` iterators with indices

### CODE REVIEW

File: Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp [hunk 1]
```diff
@@ -634,10 +634,9 @@ class CombinatorialKalmanFilter {
       // Reset the navigation state
       // Set targetSurface to nullptr for forward filtering; it's only needed
       // after smoothing
-      navigator.resetState(
-          state.navigation, state.geoContext, stepper.position(state.stepping),
-          state.options.direction * stepper.direction(state.stepping),
-          &currentState.referenceSurface(), nullptr);
```
Comment (andiwand) [inline Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:640]: this is most likely causing hash changes

File: Core/include/Acts/TrackFitting/KalmanFitter.hpp [hunk 2]
```diff
@@ -569,10 +569,8 @@ class KalmanFitter {
       // Reset navigation state
       // We do not need to specify a target here since this will be handled
       // separately in the KF actor
-      navigator.resetState(
-          state.navigation, state.geoContext, stepper.position(state.stepping),
-          state.options.direction * stepper.direction(state.stepping),
-          &st.referenceSurface(), nullptr);
```
Comment (andiwand) [inline Core/include/Acts/TrackFitting/KalmanFitter.hpp:575]: this is most likely causing hash changes

**Review by andiwand — COMMENTED** [review]
might factor that out

File: Core/include/Acts/Navigation/DetectorNavigator.hpp [hunk 3]
```diff
@@ -83,16 +83,6 @@ class DetectorNavigator {
...
 
-  void resetState(State& state, const GeometryContext& /*geoContext*/,
-                  const Vector3& /*pos*/, const Vector3& /*dir*/,
-                  const Surface* /*ssurface*/,
-                  const Surface* /*tsurface*/) const {
-    // Reset everything first
-    state = State();
-
-    // TODO fill state
-  }
```
Comment (paulgessinger) [inline Core/include/Acts/Navigation/DetectorNavigator.hpp:94]: This goes away with #2808 anyway?

File: Core/include/Acts/Propagator/StandardAborters.hpp [hunk 4]
```diff
@@ -67,15 +67,20 @@ struct PathLimitReached {
   }
 };
 
-/// This is the condition that the Surface has been reached
-/// it then triggers an propagation abort of the propagation
+/// This is the condition that the Surface has been reached it then triggers an
```
Comment (paulgessinger) [inline Core/include/Acts/Propagator/StandardAborters.hpp:70]:
```suggestion
/// This is the condition that the Surface has been reached it then triggers a
```

File: Core/include/Acts/Utilities/Intersection.hpp [hunk 5]
```diff
@@ -232,26 +233,27 @@ namespace detail {
...
 /// @param intersection The intersection to check
-/// @param pLimit The path-limit
-/// @param oLimit The overstep-limit
-/// @param tolerance The tolerance that is applied to the path-limit criterion
+/// @param nearLimit The minimum distance for an intersection to be considered
+/// @param farLimit The maximum distance for an intersection to be considered
 /// @param logger A optionally supplied logger which prints out a lot of infos
-/// at VERBOSE level
+///               at VERBOSE level
 template <typename intersection_t, typename logger_t = std::false_type>
```
Comment (paulgessinger) [inline Core/include/Acts/Utilities/Intersection.hpp:244]: You didn't touch this, but what's the `logger_t`? Can this go? I don't think it's needed.

File: Core/include/Acts/Navigation/NavigationState.hpp [hunk 6]
```diff
@@ -80,16 +82,20 @@ struct NavigationState {
 
   /// That are the candidate surfaces to process
   SurfaceCandidates surfaceCandidates = {};
-  SurfaceCandidates::const_iterator surfaceCandidate = surfaceCandidates.cend();
+  std::size_t surfaceCandidateIndex = 0;
```
Comment (paulgessinger) [inline Core/include/Acts/Navigation/NavigationState.hpp:85]: So you're switching to index based candidates here. I understand why that's needed for #2625, but outside of the scope of that rewrite, does it have any benefits over the iterators?

File: Core/src/Geometry/GenericApproachDescriptor.cpp [hunk 7]
```diff
@@ -10,6 +10,7 @@
 
 #include "Acts/Surfaces/Surface.hpp"
 #include "Acts/Utilities/Intersection.hpp"
+#include "Acts/Utilities/Logger.hpp"
```
Comment (paulgessinger) [inline Core/src/Geometry/GenericApproachDescriptor.cpp:13]: IWYU?

File: Examples/Algorithms/Fatras/src/FatrasSimulation.cpp [hunk 8]
```diff
@@ -120,15 +120,20 @@ struct FatrasSimulationT final : ActsExamples::detail::FatrasSimulation {
...
+                                           Acts::getDefaultLogger("nav", lvl)},
+                           Acts::getDefaultLogger("prop", lvl)),
+                       Acts::getDefaultLogger("Simulation", lvl)),
+                   NeutralSimulation(
+                       NeutralPropagator(
+                           NeutralStepper(),
+                           Acts::Navigator{{cfg.trackingGeometry},
+                                           Acts::getDefaultLogger("nav", lvl)},
+                           Acts::getDefaultLogger("prop", lvl)),
+                       Acts::getDefaultLogger("Simulation", lvl))) {
```
Comment (paulgessinger) [inline Examples/Algorithms/Fatras/src/FatrasSimulation.cpp:136]: Seems like an orthogonal change. Other than that, we could start with a single logger and then clone them with suffix?

File: Core/include/Acts/Navigation/DetectorNavigator.hpp (hunk 3 above)
Comment (andiwand) [reply Core/include/Acts/Navigation/DetectorNavigator.hpp:94]: looks like I forgot this one in #2808

**Comment by acts-project-service** [comment]
## :white_check_mark: Athena integration test results [930a5d78fae86c703e24c98325749be4708b8c1a]
### :white_check_mark: All tests successful

|status|job|report|
|------|---|------|
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34782085)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsBenchmarkWithSpot.sh 8 100](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34782084)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsWorkflow.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34782083)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateAmbiguityResolution.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34782082)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateResolvedTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34782081)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34782080)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsCoreSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34782079)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34782078)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34782077)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateOrthogonalSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34782076)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateClusters.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34782075)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsPersistifyEDM.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34782074)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsGSFRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34782073)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsKfRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34782072)||
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsExtrapolationAlgTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34782071)||
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsITkTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34782070)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34782069)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34782068)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34782067)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34782066)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34782065)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34782064) | [:paperclip:](https://acts-athena-outputs.web.cern.ch/test_data18_13TeV_1000evt/pulls/2768/)|
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34782063) | [:paperclip:](https://acts-athena-outputs.web.cern.ch/test_ttbarPU40_reco/pulls/2768/)|

File: Core/include/Acts/Navigation/NavigationState.hpp (hunk 6 above)
Comment (andiwand) [reply Core/include/Acts/Navigation/NavigationState.hpp:85]: what's nice about indices is that the whole state gets trivially copyable

File: Examples/Algorithms/Fatras/src/FatrasSimulation.cpp (hunk 8 above)
Comment (andiwand) [reply Examples/Algorithms/Fatras/src/FatrasSimulation.cpp:136]: right I just did not have the space for a temporary object but I think I will drop the change from here

**Review by paulgessinger — APPROVED** [review]

**Comment by paulgessinger** [comment]
Seems like there are compile errors

**Review by paulgessinger — APPROVED** [review]
